pydis_site/apps/home/signals.py

```diff
@@ -0,0 +1,159 @@
+from typing import List, Type
+
+from allauth.account.signals import user_logged_in
+from allauth.socialaccount.models import SocialAccount, SocialLogin
+from allauth.socialaccount.providers.base import Provider
+from allauth.socialaccount.providers.discord.provider import DiscordProvider
+from allauth.socialaccount.signals import (
+    pre_social_login, social_account_added, social_account_removed,
+    social_account_updated)
+from django.contrib.auth.models import Group, User as DjangoUser
+from django.db.models.signals import post_save
+
+from pydis_site.apps.api.models import User as DiscordUser
+from pydis_site.apps.staff.models import RoleMapping
+
+
+class SignalListener:
+    """
+    Listens to and processes events via the Django Signals system.
+
+    Django Signals is basically an event dispatcher. It consists of Signals (which are the events)
+    and Receivers, which listen for and handle those events. Signals are triggered by Senders,
+    which are essentially just any class at all, and Receivers can filter the Signals they listen
+    for by choosing a Sender, if required.
+
+    Signals themselves define a set of arguments that they will provide to Receivers when the
+    Signal is sent. They are always keyword arguments, and Django recommends that all Receiver
+    functions accept them as `**kwargs` (and will supposedly error if you don't do this),
+    supposedly because Signals can change in the future and your receivers should still work.
+
+    Signals do provide a list of their arguments when they're initially constructed, but this
+    is purely for documentation purposes only and Django does not enforce it.
```

Comment:
It's nice that you document this, but I think it would be better to just provide a link to the Django documentation for these three paragraphs

Reply:
I agree with you, @lemonsaurus asked me to document it though.

Reply:
There is a link to the Django docs, too. This is fine.